File: swift/Sources/ShapesBaseCommons/BaseShape.swift
import Foundation

/// BaseShape is a class that represents a generic shape with dimensions.
open class BaseShape {
    private let _dimensions: [[String: Double]]

    /// Initialize the BaseShape with an array of associative arrays
    /// representing dimensions. Each associative array should have a
    /// single key-value pair, e.g., `["radius" => 12]`.
    public init(dimensions: [[String: Double]]) {
        self._dimensions = dimensions
    }

    public var dimensions: [[String: Double]] {
        return _dimensions
    }

    public func showAvailableDimensionCount() -> Int {
        return _dimensions.count
    }

    public static func validateShape() -> Bool {
        return true
    }

    open func printDescription() {
        fatalError("Must override printDescription() in subclass")
    }
}

File: ruby/lib/shapes/base/commons/base_shape.rb
# frozen_string_literal: true

class BaseShape
    attr_reader :dimensions

    def initialize(dimensions = [])
        # Initialize the BaseShape with a list of hashes representing dimensions.
        # Each hash should have a single key-value pair, e.g., `[{"width" => 12},
        # {"length" => 15}]`.
        @dimensions = dimensions
    end

    def print_description
        raise NotImplementedError, "Must implement print_description in #{self.class}"
    end

    def show_available_dimension_count
        @dimensions.length
    end

    def self.validate_shape
        true
    end
end
